src/components/include/crt_chan.h

```diff
@@ -0,0 +1,233 @@
+ */
+
+#ifndef CRT_CHAN_H
+#define CRT_CHAN_H
+
+/***
+ *
+ */
+
+#include <cos_component.h>
+#include <crt_blkpt.h>
+#include <bitmap.h>
+
+struct crt_chan {
+	u32_t producer;
+	/* If the ring is empty, recving threads will block on this blkpt. */
+	struct crt_blkpt empty;
+	char _padding1[CACHE_LINE * 2 - (sizeof(struct crt_blkpt) + sizeof(u32_t))];
+	u32_t consumer;
+	/* If the ring is full, sending thread will block on this blkpt. */
+	struct crt_blkpt full;
+	char _padding2[CACHE_LINE * 2 - (sizeof(struct crt_blkpt) + sizeof(u32_t))];
+	/*
+	 * @item_sz is a power of two and corresponds to the
+	 * wraparound_mask. The number of data items that the channel
+	 * can hold is item_sz - 1. @wraparound_mask = nslots-1 (were
+	 * nslots is a power of two)
+	 */
+	u32_t item_sz, wraparound_mask;
+	u32_t nslots;
+	/* The memory for the channel. */
+	char mem[0];
+};
+
+/* produce a  */
+#define CRT_CHAN_STATIC_ALLOC(name, type, nslots)		\
+struct __crt_chan_envelope_##name {	                        \
+        struct crt_chan c;					\
+	char mem[nslots * sizeof(type)];			\
+} __##name;							\
+struct crt_chan *name = &__##name.c
+
+#define CRT_CHAN_TYPE_PROTOTYPES(name, type, nslots)			\
+static inline int							\
+crt_chan_init_##name(struct crt_chan *c)				\
+{ return crt_chan_init(c, sizeof(type), nslots); }			\
+static inline void							\
+crt_chan_teardown_##name(struct crt_chan *c)				\
+{ crt_chan_teardown(c); }						\
+static inline int							\
+crt_chan_empty_##name(struct crt_chan *c)				\
+{ return __crt_chan_empty(c, nslots - 1); }				\
+static inline int							\
+crt_chan_full_##name(struct crt_chan *c)				\
+{ return __crt_chan_full(c, nslots - 1); }				\
+static inline int							\
+crt_chan_send_##name(struct crt_chan *c, void *item)			\
+{									\
+	assert(pow2(nslots));						\
+	return __crt_chan_send(c, item, nslots - 1, sizeof(type));	\
+}									\
+static inline int							\
+crt_chan_recv_##name(struct crt_chan *c, void *item)			\
+{									\
+	assert(pow2(nslots));						\
+	return __crt_chan_recv(c, item, nslots - 1, sizeof(type));	\
+}									\
+static inline int							\
+crt_chan_async_send_##name(struct crt_chan *c, void *item)		\
+{									\
+	assert(pow2(nslots));						\
+	if (__crt_chan_produce(c, item, nslots - 1, sizeof(type))) return -EAGAIN; \
+	return 0;							\
+}									\
+static inline int							\
+crt_chan_async_recv_##name(struct crt_chan *c, void *item)		\
+{									\
+	assert(pow2(nslots));						\
+	if (__crt_chan_consume(c, item, nslots - 1, sizeof(type))) return -EAGAIN; \
+	return 0;							\
+}
+
+#define CRT_CHANCHAN_PROTOTYPES(nslots) \
+CRT_CHAN_TYPE_PROTOTYPES(chan, struct chan *, nslots
+
+static inline unsigned int
+__crt_chan_buff_idx(struct crt_chan *c, u32_t v, u32_t wraparound_mask)
+{ return v & wraparound_mask; }
+
+static inline int
+__crt_chan_full(struct crt_chan *c, u32_t wraparound_mask)
+{ return c->consumer == __crt_chan_buff_idx(c, c->producer + 1, wraparound_mask); }
```

Comment:
This looks like a bug... should be modified to:
**return __crt_chan_buff_idx(c, c->consumer, wraparound_mask) == __crt_chan_buff_idx(c, c->producer + 1, wraparound_mask);**
or the channel crashes after a wraparound when it is full.